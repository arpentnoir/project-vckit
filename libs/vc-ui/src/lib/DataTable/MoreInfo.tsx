import MoreVertIcon from '@mui/icons-material/MoreVert';
import { ButtonBase, Menu, MenuItem, Stack } from '@mui/material';
import { GridCellParams } from '@mui/x-data-grid';
import React from 'react';

interface IMoreInfo {
  items: { [key: string]: () => void };
  params: GridCellParams;
}

export const MoreInfo = ({ items, params }: IMoreInfo) => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    setAnchorEl(event.currentTarget);
  };

  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <Stack
      direction="row"
      style={{ justifyContent: 'end', marginLeft: 'auto' }}
    >
      <ButtonBase
        onClick={handleClick}
        data-testid={`more-info-menu-button:${params.id}`}
      >
        <MoreVertIcon />
      </ButtonBase>
      <Menu
        id={`basic-menu:${params.id}`}
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        MenuListProps={{
          'aria-labelledby': 'basic-button',
        }}
      >
        {Object.entries(items).map(([name, handleAction]) => {
          return (
            <MenuItem
              data-testid={`more-info-list-item:${name}`}
              key={name}
              onClick={handleAction}
            >
              {name}
            </MenuItem>
          );
        })}
      </Menu>
    </Stack>
  );
};
