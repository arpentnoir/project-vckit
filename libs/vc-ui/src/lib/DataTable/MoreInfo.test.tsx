import { GridCellParams } from '@mui/x-data-grid';
import { fireEvent, render } from '@testing-library/react';
import { MoreInfo } from './MoreInfo';

const mockViewFunction = jest.fn();
const mockDownloadFunction = jest.fn();
describe('MoreInfo', () => {
  it('should render correctly', () => {
    const { baseElement } = render(
      <MoreInfo
        params={{ id: 'test' } as GridCellParams}
        items={{ View: mockViewFunction, Download: mockDownloadFunction }}
      />
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should show correct items when clicked', () => {
    const { getByText, getByTestId, queryByText } = render(
      <MoreInfo
        params={{ id: 'test' } as GridCellParams}
        items={{ View: mockViewFunction, Download: mockDownloadFunction }}
      />
    );

    expect(queryByText('View')).toBeFalsy();
    expect(queryByText('Download')).toBeFalsy();

    fireEvent.click(getByTestId('more-info-menu-button:test'));

    getByText('View');
    getByText('Download');
  });

  it('should call correct function when item is clicked', () => {
    const { getByTestId } = render(
      <MoreInfo
        params={{ id: 'test' } as GridCellParams}
        items={{ View: mockViewFunction, Download: mockDownloadFunction }}
      />
    );

    fireEvent.click(getByTestId('more-info-menu-button:test'));

    fireEvent.click(getByTestId('more-info-list-item:View'));

    expect(mockViewFunction).toHaveBeenCalled();

    fireEvent.click(getByTestId('more-info-list-item:Download'));

    expect(mockDownloadFunction).toHaveBeenCalled();
  });
});
