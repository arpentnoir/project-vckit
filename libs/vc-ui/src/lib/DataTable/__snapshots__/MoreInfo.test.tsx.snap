// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoreInfo should render correctly 1`] = `
<body>
  <div>
    <div
      class="css-m69qwo-MuiStack-root"
      style="justify-content: end; margin-left: auto;"
    >
      <button
        class="MuiButtonBase-root css-10d1a0h-MuiButtonBase-root"
        data-testid="more-info-menu-button:test"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MoreVertIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;
